SCClassLibrary/Common/Streams/EventStreamCleanup.sc

```diff
@@ -1,47 +1,86 @@
+// The "new" cleanup framework: built-in mechanism to avoid duplicate
+// execution of cleanups, difficult to avoid otherwise due to their
+// replicated nature in a stream graph. Instead of storing bare Functions,
+// they are now stored in CleanupThunks.
+
+// Unlike Thunks which take no arguments, CleanupThunks do allow arguments
+// to be passed to the function, but the CleanupThunk contract is that any
+// subsequent calls into the function would monotonically do the same or
+// less (cleanup) work, so that the first call does the maximal amount of
+// relevant work, making subsequent calls completely redundant.
+// Prototypical example: if a cleanup is called with a "flag: false",
+// it is assumed it will *not* be called later again with "flag: true",
+// if the latter means that the function would have to do "extra" work
+// that was skipped/excluded by "flag: false".
+// CleanupThunk is not a general-purpose function memoization class for
+// functions with arguments.
+CleanupThunk : Thunk {
+
+	value { |... args| ^this.prEvaluate(\valueArray, args) }
+	valueArray { |... args| ^this.prEvaluate(\valueArray, *args) }
+	valueEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, args) }
+	valueArrayEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, *args) }
+
+	prEvaluate { |selector, args|
+		^value ?? {
+			value = function.performList(selector, args);
+			function = nil;
+			value
+		}
+	}
+	didEvaluate { ^function.isNil }
+	clear { function = nil }
+}
```

Comment:
It would be reasonable to move this up as well. But perhaps `cancel` may be a better name, because it doesn't interfere with the existing behaviour.

Reply:
Moving this up, but we have a `clear` protocol already. I think we should stick with it.